internal/services/circle_reconciliation_service.go

```diff
@@ -61,68 +70,108 @@ func (s *CircleReconciliationService) Reconcile(ctx context.Context) error {
-			if newStatus.IsCompleted() {
-				completedAt = &now
-			}
-			circleRequest, err = s.Models.CircleTransferRequests.Update(ctx, dbTx, circleRequest.IdempotencyKey, data.CircleTransferRequestUpdate{
-				Status:            newStatus,
-				CompletedAt:       completedAt,
-				LastSyncAttemptAt: &now,
-				SyncAttempts:      circleRequest.SyncAttempts + 1,
-				ResponseBody:      jsonBody,
-			})
-			if err != nil {
-				return fmt.Errorf("updating Circle transfer request: %w", err)
-			}
+	if len(reconciliationErrors) > 0 {
+		return fmt.Errorf("attempted to reconcyle %d circle requests but failed on %d reconciliations: %v", reconciliationCount, len(reconciliationErrors), reconciliationErrors)
+	}
 
-			// 4.3. update the payment status in the DB.
-			newPaymentStatus, err := transfer.Status.ToPaymentStatus()
-			if err != nil {
-				return fmt.Errorf("converting Circle transfer status to Payment status: %w", err)
-			}
-			var statusMsg string
-			switch newStatus {
-			case data.CircleTransferStatusSuccess:
-				statusMsg = fmt.Sprintf("Circle transfer completed successfully with the Stellar transaction hash: %q", transfer.TransactionHash)
-			case data.CircleTransferStatusFailed:
-				statusMsg = fmt.Sprintf("Circle transfer failed with error: %q", transfer.ErrorCode)
-			default:
-				return fmt.Errorf("unexpected Circle transfer status: %q", newStatus)
-			}
+	return nil
+}
 
-			err = s.Models.Payment.UpdateStatus(ctx, dbTx, circleRequest.PaymentID, newPaymentStatus, &statusMsg, transfer.TransactionHash)
-			if err != nil {
-				return fmt.Errorf("updating payment status: %w", err)
-			}
+// shouldIncrementSyncAttempts returns true if the error is a bad request error from Circle API, which means that the
+// request should be retried.
+func shouldIncrementSyncAttempts(err error) bool {
+	var cAPIErr *circle.APIError
+	if !errors.As(err, &cAPIErr) {
+		return false
+	}
 
-			log.Ctx(ctx).Infof("[tenant=%s] Reconciled Circle transfer request %q with status %q", tnt.Name, *circleRequest.CircleTransferID, newStatus)
+	return cAPIErr.StatusCode == http.StatusBadRequest
+}
+
+// reconcileTransferRequest reconciles a Circle transfer request and updates the payment status in the DB. It returns an
+// error if the reconciliation fails.
+func (s *CircleReconciliationService) reconcileTransferRequest(ctx context.Context, dbTx db.DBTransaction, tnt *tenant.Tenant, circleRequest *data.CircleTransferRequest) error {
+	// 4.1. get the Circle transfer by ID
+	transfer, err := s.CircleService.GetTransferByID(ctx, *circleRequest.CircleTransferID)
+	if err != nil {
+		if shouldIncrementSyncAttempts(err) {
+			// increment the sync attempts and update the last sync attempt time.
+			var updateErr error
+			circleRequest, updateErr = s.Models.CircleTransferRequests.Update(ctx, dbTx, circleRequest.IdempotencyKey, data.CircleTransferRequestUpdate{
+				LastSyncAttemptAt: utils.TimePtr(time.Now()),
+				SyncAttempts:      circleRequest.SyncAttempts + 1,
```

Comment:
In case where we're incrementing the sync attempt, we should also be having a clear error message from circle that we can persist with this request as well. 
Maybe we can add a TODO to persist that later as well ?

Reply:
Good point, I can store it in the ResponseBody